tony-history-server/app/utils/HdfsUtils.java

```diff
@@ -0,0 +1,108 @@
+    try {
+      return fs.exists(filePath);
+    } catch (IOException e) {
+      LOG.error(filePath.toString() + " doesn't exist!");
+      return false;
+    }
+  }
+
+  static String contentOfHdfsFile(FileSystem fs, Path filePath) {
+    StringBuilder fileContent = new StringBuilder();
+    try (FSDataInputStream inStrm = fs.open(filePath);
+        BufferedReader bufReader = new BufferedReader(new InputStreamReader(inStrm))) {
+      String line;
+      while ((line = bufReader.readLine()) != null) {
+        fileContent.append(line);
+      }
+      return fileContent.toString();
+    } catch (IOException e) {
+      LOG.error("Couldn't read content of file from HDFS", e);
+      return "";
+    }
+  }
+
+  static String copyFromHdfs(FileSystem fs, Path hdfsFilePath) {
+    String localFileName = "config.xml";
+    try (FSDataInputStream inStrm = fs.open(hdfsFilePath); FileWriter fw = new FileWriter(localFileName)) {
+      IOUtils.copy(new InputStreamReader(inStrm), fw);
+    } catch (IOException e) {
+      LOG.error("Couldn't copy from file from HDFS", e);
+      return "";
+    }
+    return localFileName;
+  }
+
+  static List<Path> getValidPaths(FileStatus[] lsJobDir, Predicate<FileStatus> fn) {
+    return Arrays.stream(lsJobDir)
+        .filter(fileStatus -> fn.test(fileStatus))
+        .map(fileStatus -> fileStatus.getPath())
+        .collect(Collectors.toList());
+  }
+
+  public static List<Path> getFilePathsFromAllJobs(FileSystem fs, String histFolder, String fileType) {
+    List<Path> paths = new ArrayList<>();
+    FileStatus[] lsHist;
+    try {
+      lsHist = fs.listStatus(new Path(histFolder));
+    } catch (Exception e) {
+      LOG.error("Failed scanning " + histFolder, e);
+      return paths;
+    }
+
+    LOG.debug("lsHist size: " + lsHist.length);
+    paths = Arrays.stream(lsHist).filter(fileStatus -> fileStatus.isDirectory()).map((item) -> {
+      try {
+        return getValidPaths(fs.listStatus(new Path(item.getPath().toString())),
+            fileStatus -> fileStatus.getPath().toString().contains(fileType));
+      } catch (Exception e) {
+        LOG.error("Failed scanning " + item.getPath().toString(), e);
+        return new ArrayList<Path>();
+      }
+    }).flatMap(List::stream).collect(Collectors.toList());
+    return paths;
+  }
+
+  public static List<Path> getFilePathsFromOneJob(FileSystem fs, String histFolder, String jobId, String fileType) {
+    List<Path> paths = new ArrayList<>();
+    StringBuilder jobDirSb = new StringBuilder();
+    jobDirSb.append(histFolder);
+    jobDirSb.append(jobId);
+    jobDirSb.append("/");
+    FileStatus[] lsJobDir;
+
+    try {
+      lsJobDir = fs.listStatus(new Path(jobDirSb.toString()));
+    } catch (Exception e) {
+      LOG.error("Failed scanning " + jobDirSb.toString());
+      return paths;
+    }
+
+    return getValidPaths(lsJobDir, fileStatus -> fileStatus.getPath().toString().contains(fileType));
```

Comment:
Instead of implementing your own fileType filter, you can just use `FileSystem.listStatus(Path f, PathFilter filter)`.

Reply:
I'd still wrap it around a helper function though, since we still need to convert `FileStatus[]` to `List<Path>`. I'm not sure if it helps in terms of readability and/or maintainability. Here's a gist after refactoring: https://gist.github.com/pdtran3k6/9003ae5235342541259e36c7ccef7aa4

Reply:
Ok, fine with using our own predicate if you think it's more readable.

Suggest you change the `.contains(fileType)` to `endsWith(fileType)` to be more robust.

Reply:
👍